### DISCUSSION TITLE
Unregister ros1 subscribers?

### DISCUSSION THREAD

**Discussion by lucasw** [RosLibRust/roslibrust#200] [Q&A]
Is there a current method for shutting down subscribers fully?  So far I'm having luck unregistering publishers, if I make sure they go out of scope in their tokio tasks before the process exits then they don't appear in rosnode info for the dead node, but subscribers linger and a rosnode cleanup is needed.

I see `unregister_subscriber()` in https://github.com/Carter12s/roslibrust/blob/master/roslibrust/src/ros1/master_client.rs#L224 - but it's not exposed through a node handle, nothing calls it?

Beyond that what about interrupting `my_subscriber.next().await`- I have several of those running inside tokio tasks- if unregister_subscriber were exposed would/could it cause that next() to return an error, like how a channel errors if the other end of it (sender or receiver) goes away?

**Comment by lucasw** [comment]
I have a manual unregister_subscriber working in https://github.com/lucasw/roslibrust/tree/get_rx_msg_def https://github.com/lucasw/roslibrust/commit/5fbbd2ef0515f208b56fb2c6ad2efd68a60b16ba - but so far I haven't been able to get `drop` to do it automatically.

I'll rebase the minimal version of it onto master and make at least a draft PR.

**Reply by Carter12s** [reply]
Nice!

I had the chance to look at your code this morning, and then while I was on a flight today decided to implement this:
https://github.com/Carter12s/roslibrust/pull/204

It isn't a full answer to your question and we should make Subscriptions clean up after themselves, but I added a Node level cleanup of any ROS references that (at least in my testing) means we don't leave any junk in `rosnode info` / `rosnode list`.

**Reply by Carter12s** [reply]
I'd love to see a draft PR of having Subscribers cleanup after themselves. I think the best paradigm for how to do this right is what ServiceServer does. I think having each of Publisher, Subscriber, and ServiceServer hold a weak_clone() of NodeHandle and then to call a non-async clean-up function on NodeHandle in their Drop impl will end up working well and being pretty readable.

**Reply by DomWilliamsEE** [reply]
Is this still the best approach? I am experimentally porting a node from rosrust but unsubscribing from expensive topics on the fly is a requirement.

**Comment by Carter12s** [comment]
Fixed by @DomWilliamsEE in #323
